src/sources/dnstap/mod.rs

```diff
@@ -0,0 +1,509 @@
+            max_frame_handling_tasks,
+            socket_file_mode,
+            socket_receive_buffer_size,
+            socket_send_buffer_size,
+            host_key,
+            timestamp_key: timestamp_key.to_string(),
+        }
+    }
+}
+
+impl FrameHandler for DnstapFrameHandler {
+    fn content_type(&self) -> String {
+        self.content_type.clone()
+    }
+
+    fn max_frame_length(&self) -> usize {
+        self.max_frame_length
+    }
+
+    /**
+     * Function to pass into util::framestream::build_framestream_unix_source
+     * Takes a data frame from the unix socket and turns it into a Vector Event.
+     **/
+    fn handle_event(&self, received_from: Option<Bytes>, frame: Bytes) -> Option<Event> {
+        let mut event = Event::new_empty_log();
+
+        let log_event = event.as_mut_log();
+
+        let frame_size = frame.len();
+
+        if let Some(host) = received_from {
+            log_event.insert(self.host_key(), host);
+        }
+
+        if self.raw_data_only {
+            log_event.insert(
+                &self.schema.dnstap_root_data_schema().raw_data(),
+                base64::encode(&frame),
+            );
+            emit!(DnstapEventReceived {
+                byte_size: frame_size
+            });
+            Some(event)
+        } else {
+            match DnstapParser::new(&self.schema, log_event).parse_dnstap_data(frame) {
+                Err(err) => {
+                    emit!(DnstapParseDataError {
+                        error: format!("Dnstap protobuf decode error {:?}.", err).as_str()
+                    });
+                    None
+                }
+                Ok(_) => {
+                    emit!(DnstapEventReceived {
+                        byte_size: frame_size
+                    });
+                    Some(event)
+                }
+            }
+        }
+    }
+    fn socket_path(&self) -> PathBuf {
+        self.socket_path.clone()
+    }
+
+    fn multithreaded(&self) -> bool {
+        self.multithreaded
+    }
+
+    fn max_frame_handling_tasks(&self) -> u32 {
+        self.max_frame_handling_tasks
+    }
+
+    fn socket_file_mode(&self) -> Option<u32> {
+        self.socket_file_mode
+    }
+
+    fn socket_receive_buffer_size(&self) -> Option<usize> {
+        self.socket_receive_buffer_size
+    }
+
+    fn socket_send_buffer_size(&self) -> Option<usize> {
+        self.socket_send_buffer_size
+    }
+
+    fn host_key(&self) -> String {
+        self.host_key.clone()
+    }
+
+    fn timestamp_key(&self) -> String {
+        self.timestamp_key.clone()
+    }
+}
+
+#[cfg(all(test, feature = "dnstap-integration-tests"))]
+mod integration_tests {
+    use super::*;
+    use crate::{event::Value, test_util::trace_init, Pipeline};
+    use futures::StreamExt;
+    use serde_json::json;
+    use std::{env, path::Path, process::Command, thread};
+    use tokio::time;
+
+    async fn test_dnstap(raw_data: bool, query_type: &'static str) {
```

Comment:
> So `raw_message` is expanded out into `raw_message_clone`? That is, that they end up being the same, just that raw_message_clone has the domain names uncompressed? If so, I think we could call this uncompressed_raw_message.

Well, no exactly. Nothing in `raw_message_clone` is uncompressed and altered in place. If there are compressed domain names in `raw_message`, they'll remain "compressed" in `raw_message_clone`. The `raw_message_clone` is basically used as destinations referenced by pointers contained in compressed domain names. See section "4.4.4. Message compression" of [RFC 1035 ](https://tools.ietf.org/html/rfc1035) for the domain name compression algorithm.

Reply:
Ah, I see. Why does it need to be longer than? It seems like it'd be the same as `raw_message`.

Reply:
Good question. As RFC 1035 describes: a compressed domain name may replace labels in the domain name with pointers. These pointers are basically offset from the very beginning of the raw DNS message. By following these pointers, we'd be able to reconstruct the entire domain name. However, when `trustdns` encounters an unsupported rdata, it'll return only the `rdata` content through `rdata::NULL`, which cannot be used directly by `trustdns` method `Name::read()` to uncompress domain names, as the pointers inside the domain name may reference to somewhere outside of `rdata`.
We could try to find the `rdata` in `raw_message` and then do the uncompression there, but it'd be inefficient to do so, especially when the message is large and the rdata locates at the end of the message. To optimized the process, we introduced in the clone of `raw_message` but with doubled capacity, so that we can simply append the `rdata` to the end of it, and pass the entire `raw_message` clone to method `Name::read()` (inside a decoder) to reconstruct the name.